lib/github.dart

```diff
@@ -0,0 +1,1029 @@
+  Stream get onStarredGistListChanged =>
+      _starredGistListUpdateController.stream;
+  Stream get onGistStarredCheckerReport =>
+      _gistStarredCheckerReportController.stream;
+
+  final List<Gist> _myGistList = [];
+  final List<Gist> _starredGistList = [];
+
+  List<Gist> get myGistList => _myGistList;
+  List<Gist> get starredGistList => _starredGistList;
+
+  String? _pendingUserInfoRequest;
+  String? _pendingUserGistRequest;
+  String? _pendingUserStarredGistRequest;
+
+  GitHubAuthenticationController(this.launchUri, this.snackbar,
+      {http.Client? client}) {
+    // Check for parameters in query uri.
+    _client = client ?? http.Client();
+
+    final params = Map<String, String?>.from(launchUri.queryParameters);
+    final String ghTokenFromUrl = params['gh'] ?? '';
+    final String ghScope = params['scope'] ?? '';
+
+    if (ghTokenFromUrl.isNotEmpty) {
+      final String perAuthParamsJson =
+          window.localStorage['gh_pre_auth_query_params'] ?? '';
+
+      try {
+        final restoreParams = Map<String, String?>.from(
+            json.decode(perAuthParamsJson) as Map<dynamic, dynamic>);
+
+        final Uri restoredUrl =
+            launchUri.replace(queryParameters: restoreParams);
+        window.history.replaceState({}, 'DartPad', restoredUrl.toString());
+      } catch (e) {
+        window.console
+            .log('Caught doing restoreParams exception ${e.toString()}');
+      }
+
+      if (ghTokenFromUrl == 'noauth' || ghTokenFromUrl == 'authfailed') {
+        // ERROR was encountered during trip to GH auth.
+        snackbar.showMessage('Error encountered during GitHub OAuth Request');
+        return;
+      }
+      if (!ghScope.contains('gists')) {
+        // Give error message but continue in this case.
+        snackbar.showMessage(
+            'Error: The scope "gists" was not included with the GitHub OAuth Token');
+      }
+
+      // Now decrypt the GH token and try and init user.
+      final String ghAuthToken =
+          decryptAuthTokenFromReturnedSecureAuthToken(ghTokenFromUrl);
+
+      // Set provided a gh token, if new this will do query on user info.
+      githubOAuthAccessToken = ghAuthToken;
+    } else {
+      // There was no gh token in the window URL, but we may have STORED GH
+      // authorization in local storage...
+      // so trigger an authentication state change anyway
+      //window.console.log(
+      //    'No GH query param but do we have a STORED local storage token ?');
+      //window.console.log('githubOAuthAccessToken = $githubOAuthAccessToken');
+    }
+  }
+
+  void postCreationFireAutheticatedStateChangeEvent() {
+    _authenticatedStateChangeController.add(githubOAuthAccessToken != '');
+    updateUsersGistAndStarredGistsList();
+  }
+
+  Timer? starGistsChecklDelayTimer;
+
+  void updateUsersGistAndStarredGistsList({int starredCheckDelay = 100}) {
+    // Now go and get the lists of user's gists and starred gists.
+    getUsersGists();
```

Comment:
What's the point of `starredCheckDelay`? It might be better to use async/await with a [CancelableCompleter](https://pub.dev/documentation/async/latest/async/CancelableCompleter-class.html)

Reply:
The UI always reflects what the GitHub api is currently returning.
After the user initiates an operation to star a gist the GitHub api can take some time to return the newly starred gist as being starred..  Because of this - after the user chooses to star a gist this routine can be told to wait X milliseconds before checking with the GitHub api for the list of starred gists, and then updating the UI accordingly (and hopefully the GitHub api will return the starred list that the user expects after the star/unstar operation they initiated.)
I chose this as the simplest of other more complex approaches that might be taken to handle this problem.